engine/src/main/java/org/pentaho/di/repository/BaseRepositoryMeta.java

```diff
@@ -0,0 +1,244 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2017 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.repository;
+
+import java.util.List;
+import java.util.Map;
+
+import org.json.simple.JSONObject;
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.database.DatabaseMeta;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.w3c.dom.Node;
+
+public class BaseRepositoryMeta {
+
+  public static final String ID = "id";
+  public static final String DISPLAY_NAME = "displayName";
+  public static final String DESCRIPTION = "description";
+  public static final String IS_DEFAULT = "isDefault";
+
+  protected String id;
+  protected String name;
+  protected String description;
+  protected Boolean isDefault = false;
+
+  public BaseRepositoryMeta( String id ) {
+    this.id = id;
+  }
+
+  /**
+   * This returns the expected name for the dialog that edits this repository metadata object The expected name is in
+   * the org.pentaho.di.ui tree and has a class name that is the name of the job entry with 'Dialog' added to the end.
+   *
+   * e.g. if the repository meta class is org.pentaho.di.repository.kdr.KettleDatabaseRepositoryMeta the dialog would be
+   * org.pentaho.di.ui.repository.kdr.KettleDatabaseRepositoryDialog
+   *
+   * If the dialog class does not match this pattern, the RepositoryMeta class should override this method and return
+   * the appropriate class name
+   *
+   * @return full class name of the dialog
+   */
+  public String getDialogClassName() {
+    String className = getClass().getCanonicalName();
+    className = className.replaceFirst( "\\.di\\.", ".di.ui." );
+    if ( className.endsWith( "Meta" ) ) {
+      className = className.substring( 0, className.length() - 4 );
+    }
+    className += "Dialog";
+    return className;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 64:70, 86:92.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.